cuda/solver/gmres_kernels.cu

```diff
@@ -48,63 +57,687 @@ namespace cuda {
+{
+    const auto tidx = threadIdx.x;
+    const auto col_idx = blockIdx.x;
+
+    // Used that way to get around dynamic initialization warning and
+    // template error when using `reduction_helper_array` directly in `reduce`
+    __shared__ UninitializedArray<ValueType, block_size> reduction_helper_array;
+    ValueType *__restrict__ reduction_helper = reduction_helper_array;
+
+    if (col_idx < num_cols && !stop_status[col_idx].has_stopped()) {
+        ValueType local_res{};
+        for (size_type i = tidx; i < num_rows; i += block_size) {
+            const auto next_krylov_idx = i * stride_next_krylov + col_idx;
+            const auto next_krylov_value = next_krylov_basis[next_krylov_idx];
+
+            local_res += next_krylov_value * next_krylov_value;
+        }
+
+        reduction_helper[tidx] = local_res;
+
+        // Perform thread block reduction. Result is in reduction_helper[0]
+        reduce(group::this_thread_block(), reduction_helper,
+               [](const ValueType &a, const ValueType &b) { return a + b; });
+
+        if (tidx == 0) {
+            hessenberg_iter[(iter + 1) * stride_hessenberg + col_idx] =
+                sqrt(reduction_helper[0]);
+        }
+    }
+}
+
+
+// Must be called with at least `num_rows * stride_next_krylov` threads in
+// total.
+template <typename ValueType>
+__global__
+    __launch_bounds__(default_block_size) void update_krylov_next_krylov_kernel(
+        size_type iter, size_type num_rows, size_type num_cols,
+        ValueType *__restrict__ next_krylov_basis, size_type stride_next_krylov,
+        ValueType *__restrict__ krylov_bases, size_type stride_krylov,
+        const ValueType *__restrict__ hessenberg_iter,
+        size_type stride_hessenberg,
+        const stopping_status *__restrict__ stop_status)
+{
+    const auto global_id = threadIdx.x + blockIdx.x * blockDim.x;
+    const auto row_idx = global_id / stride_next_krylov;
+    const auto col_idx = global_id % stride_next_krylov;
+    const auto hessenberg =
+        hessenberg_iter[(iter + 1) * stride_hessenberg + col_idx];
+
+    if (row_idx < num_rows && col_idx < num_cols &&
+        !stop_status[col_idx].has_stopped()) {
+        const auto next_krylov_idx = row_idx * stride_next_krylov + col_idx;
+        const auto krylov_idx =
+            row_idx * stride_krylov + num_cols * (iter + 1) + col_idx;
+
+        const auto next_krylov_value =
+            next_krylov_basis[next_krylov_idx] / hessenberg;
+
+        next_krylov_basis[next_krylov_idx] = next_krylov_value;
+        krylov_bases[krylov_idx] = next_krylov_value;
+    }
+}
+
+
+template <typename ValueType>
+void finish_arnoldi(std::shared_ptr<const CudaExecutor> exec,
+                    matrix::Dense<ValueType> *next_krylov_basis,
+                    matrix::Dense<ValueType> *krylov_bases,
+                    matrix::Dense<ValueType> *hessenberg_iter,
+                    const size_type iter, const stopping_status *stop_status)
```

Comment:
This `size_type` is const but pretty much all the other ones before are not. Is there a reason? I think we should unify to whether putting them all as `const` or none.

Reply:
No idea to be honest. It was also present in the `step_1` and `step_2` kernel definitions in `core/solver/gmres_kernels.hpp` and I was unsure if I should change that as well, so I ignored it.

I removed it in the `core/solver/gmres_kernels.hpp` because these `const` are not part of the signature.
I also removed the other `const`s (if I would make all the other const, the functions would be even more bloated) in the current version.